File: app/src/main/java/awais/instagrabber/utils/NumberUtils.java
package awais.instagrabber.utils;

import androidx.annotation.NonNull;
import androidx.core.util.Pair;

import java.util.Random;

public final class NumberUtils {
    @NonNull
    public static String millisToString(final long timeMs) {
        final long totalSeconds = timeMs / 1000;

        final long seconds = totalSeconds % 60;
        final long minutes = totalSeconds / 60 % 60;
        final long hours = totalSeconds / 3600;

        final String strSec = Long.toString(seconds);
        final String strMin = Long.toString(minutes);

        final String strRetSec = strSec.length() > 1 ? strSec : "0" + seconds;
        final String strRetMin = strMin.length() > 1 ? strMin : "0" + minutes;

        final String retMinSec = strRetMin + ':' + strRetSec;

        if (hours > 0)
            return Long.toString(hours) + ':' + retMinSec;
        return retMinSec;
    }

    public static int getResultingHeight(final int requiredWidth, final int height, final int width) {
        return requiredWidth * height / width;
    }

    public static int getResultingWidth(final int requiredHeight, final int height, final int width) {
        return requiredHeight * width / height;
    }

    public static long random(long origin, long bound) {
        final Random random = new Random();
        long r = random.nextLong();
        long n = bound - origin, m = n - 1;
        if ((n & m) == 0L)  // power of two
            r = (r & m) + origin;
        else if (n > 0L) {  // reject over-represented candidates
            //noinspection StatementWithEmptyBody
            for (long u = r >>> 1;            // ensure non-negative
                 u + m - (r = u % n) < 0L;    // rejection check
                 u = random.nextLong() >>> 1) // retry
                ;
            r += origin;
        } else {              // range not representable as long
            while (r < origin || r >= bound)
                r = random.nextLong();
        }
        return r;
    }

    @NonNull
    public static Pair<Integer, Integer> calculateWidthHeight(final int height, final int width, final int maxHeight, final int maxWidth) {
        int tempWidth = width;
        int tempHeight = height > maxHeight ? maxHeight : height;
        if (tempWidth > maxWidth) {
            tempHeight = NumberUtils.getResultingHeight(maxWidth, height, width);
            tempWidth = maxWidth;
        }
        return new Pair<>(tempWidth, tempHeight);
    }
}

File: app/src/main/java/awais/instagrabber/utils/NumberUtils.kt
@file:JvmName("NumberUtils")

package awais.instagrabber.utils

import java.util.*
import kotlin.math.ln
import kotlin.math.pow

fun getResultingHeight(requiredWidth: Int, height: Int, width: Int): Int {
    return requiredWidth * height / width
}

fun getResultingWidth(requiredHeight: Int, height: Int, width: Int): Int {
    return requiredHeight * width / height
}

// TODO Replace all usages with kotlin Random.nextLong() once converted to kotlin
fun random(origin: Long, bound: Long): Long {
    val random = Random()
    var r = random.nextLong()
    val n = bound - origin
    val m = n - 1
    when {
        n and m == 0L -> r = (r and m) + origin // power of two
        n > 0L -> {
            // reject over-represented candidates
            var u = r ushr 1 // ensure non-negative
            while (u + m - u % n.also { r = it } < 0L) { // rejection check
                // retry
                u = random.nextLong() ushr 1
            }
            r += origin
        }
        else -> {
            // range not representable as long
            while (r < origin || r >= bound) r = random.nextLong()
        }
    }
    return r
}

fun calculateWidthHeight(height: Int, width: Int, maxHeight: Int, maxWidth: Int): NullSafePair<Int, Int> {
    if (width > maxWidth) {
        var tempHeight = getResultingHeight(maxWidth, height, width)
        var tempWidth = maxWidth
        if (tempHeight > maxHeight) {
            tempWidth = getResultingWidth(maxHeight, tempHeight, tempWidth)
            tempHeight = maxHeight
        }
        return NullSafePair(tempWidth, tempHeight)
    }
    if (height < maxHeight && width < maxWidth || height > maxHeight) {
        var tempWidth = getResultingWidth(maxHeight, height, width)
        var tempHeight = maxHeight
        if (tempWidth > maxWidth) {
            tempHeight = getResultingHeight(maxWidth, tempHeight, tempWidth)
            tempWidth = maxWidth
        }
        return NullSafePair(tempWidth, tempHeight)
    }
    return NullSafePair(width, height)
}

fun roundFloat2Decimals(value: Float): Float {
    return ((value + (if (value >= 0) 1 else -1) * 0.005f) * 100).toInt() / 100f
}

fun abbreviate(number: Long, options: AbbreviateOptions? = null): String {
    // adapted from https://stackoverflow.com/a/9769590/1436766
    var threshold = 1000
    var addSpace = false
    if (options != null) {
        threshold = options.threshold
        addSpace = options.addSpaceBeforePrefix
    }
    if (number < threshold) return "" + number
    val exp = (ln(number.toDouble()) / ln(threshold.toDouble())).toInt()
    return String.format(
        Locale.US,
        "%.1f%s%c",
        number / threshold.toDouble().pow(exp.toDouble()),
        if (addSpace) " " else "",
        "kMGTPE"[exp - 1]
    )
}

data class AbbreviateOptions(val threshold: Int = 1000, val addSpaceBeforePrefix: Boolean = false)